processor/deltatocumulativeprocessor/internal/delta/delta.go

```diff
@@ -44,8 +44,17 @@ func (a Accumulator[D]) Store(id streams.Ident, dp D) error {
 		return ErrOutOfOrder{Last: aggr.Timestamp(), Sample: dp.Timestamp()}
 	}
 
+	// detect gaps
+	var gap error
+	if dp.StartTimestamp() > aggr.Timestamp() {
```

Comment:
I respectfully disagree :slightly_smiling_face: 

I am aiming for highly (mentally) decoupled components, which are understandable at a glance and composable to achieve the larger picture.

My thinking is that from the perspective of the Accumulator, it tried to accumulate a new sample. It was successful, but not perfectly so as the accumulation is now lossy. This is an error. Not a terrible one, but it's also not `nil`, which reads as "could not have gone better" to me.
In consequence, that information is returned back to the previous layer, which either does the same or has enough context to deal with that appropriately. 

In our case it appears reasonable to record non-fatal events as metrics and proceed as usual.
Having all such error (or event) handling on a higher level also makes it easy to vet that decision making, as everything that can go wrong is in one place.

Reply:
> I respectfully disagree

That's fair. It's totally fine for us to have different opinions :)

In general, I agree wholeheartedly about having separation of concerns and composability. My concern with the current setup is the amount of overloading / nesting. IE, at a glance, it's very difficult to know "who's" .Store() we're calling. And thus what it will do. You're almost forced to debug it at runtime to know what the implementation may do.

At the interface level, we're saying, "store this value please", but down the layers were doing much much more than that.


> Having all such error (or event) handling on a higher level also makes it easy to vet that decision making, as everything that can go wrong is in one place.

I agree. But I just disagree about how to do it :P

My vote would be to move more of the logic up. So everything is in (mostly) one layer. Potentially using functions to encapsulate logic. But less nesting.

So there would be a root `Store(ctx context.Context, id streams.Identifier, v T, m *metrics.Metrics)`, which would have a series of if statements to call the various logic. (Is eviction enabled, if so, check the length and evict if needed. Record appropriate metrics. Etc)


But in the end, I also recognize that this is just my opinion, and you're doing all the implementation work. :)

Reply:
That's a very valid observation, heavily leaning on composed nesting obscures how things interact together.
I think we could certainly do better here.

I tried to mitigate that a little by making the constructor very linear:
https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31363/files#diff-edcf9376e4ce5f82d566887ea227f4bbf70f6ea4edf65eba69f69a0e14168173L48-L57

For the moment however, I don't think the benefits justify the time-cost of coming up with a better design.
What do you think about coming back to this before declaring it stable (or beta if you wish)?